pkg/reconciler/controllers/syncer.go

```diff
@@ -0,0 +1,343 @@
+					continue
+				}
+				timerEvents <- timerEvent(trigger)
+
+			// Update the sync status to report management conflicts (from the remediator).
+			case <-s.statusUpdateTimer.C:
+				timerEvents <- timerEvent(parse.StatusUpdate)
+			}
+		}
+	}()
+
+	ctrlBuilder := ctrl.NewControllerManagedBy(mgr).
+		Named(SyncerController).
+		// watch for the timer events
+		Watches(&source.Channel{Source: timerEvents},
+			&handler.EnqueueRequestForObject{})
+
+	return ctrlBuilder.Complete(s)
+}
+
+func timerEvent(trigger parse.Trigger) event.GenericEvent {
+	// Enqueue the request as a Kubernetes core Event
+	req := corev1.Event{}
+	req.Name = string(trigger)
+	// Send the Event to the event channel to trigger a new reconciliation.
+	return event.GenericEvent{Object: &req}
+}
+
+// retry re-runs the parse-apply-watch loop for the given retry trigger (managementConflict, retry, or watchUpdate).
+// If resetCache is true, it clears the cache to force the parse-apply-watch loop to run.
+// In the end, it sets the next retryTimer with exponential backoff.
+// If it reaches retry, the retryTimer won't be reset, so no more retries will be scheduled.
+func (s *Syncer) retry(trigger parse.Trigger, resetCache bool) {
+	if s.state.ReachRetryLimit() {
+		klog.Infof("Retry limit (%d) has been reached. No more retries will be scheduled for the same commit.", parse.RetryLimit)
+		// Don't reset retryTimer if retry limit has been reached.
+		// retryBackoff and retryTimer will be reset when a new commit is detected.
+	} else {
+		retryDuration := s.state.NextRetryDuration()
+		retries := parse.RetryLimit - s.state.RetrySteps()
+		klog.Infof("A retry is triggered (trigger: %s, retries: %d/%d), next retry will be scheduled in %s",
+			trigger, retries, parse.RetryLimit, retryDuration)
+		// Reset the cache partially to make sure all the steps of a parse-apply-watch loop will run.
+		// The cached sourceState will not be reset to avoid reading all the source files unnecessarily.
+		// The cached needToRetry will not be reset to avoid resetting the backoff retries.
+		if resetCache {
+			s.state.ResetPartialCache()
+		}
+		parse.Run(s.controllerCtx, s.parser, trigger, s.state)
+		s.retryTimer.Reset(retryDuration) // Schedule retry attempt
+	}
+}
+
+// StartSyncing marks the Syncer is working.
+func (s *Syncer) StartSyncing() bool {
+	s.lifecycleMux.Lock()
+	defer s.lifecycleMux.Unlock()
+
+	if s.canSync {
+		klog.V(1).Infof("Start syncing...")
+		s.wg.Add(1)
+	}
+	return s.canSync
+}
+
+// WaitSyncing blocks until the Syncer finishes processing the requests.
+func (s *Syncer) WaitSyncing() {
+	klog.V(1).Info("Waiting for the Syncer to complete...")
+	s.wg.Wait()
+}
+
+// DoneSyncing marks the Syncer as done when it finishes syncing.
+func (s *Syncer) DoneSyncing() {
+	klog.V(1).Info("Syncer completes...")
+	s.wg.Done()
+}
+
+// PauseSyncing stops the Syncer to process new requests.
+func (s *Syncer) PauseSyncing() {
```

Comment:
It doesn't looks like we ever really need to resume syncing. I think this could be replaced with a Context and a DoneCh, like you have in the remediator.